staging/src/k8s.io/client-go/tools/record/event.go

```diff
@@ -343,10 +340,6 @@ func (recorder *recorderImpl) Eventf(object runtime.Object, eventtype, reason, m
 	recorder.Event(object, eventtype, reason, fmt.Sprintf(messageFmt, args...))
 }
 
-func (recorder *recorderImpl) PastEventf(object runtime.Object, timestamp metav1.Time, eventtype, reason, messageFmt string, args ...interface{}) {
-	recorder.generateEvent(object, nil, timestamp, eventtype, reason, fmt.Sprintf(messageFmt, args...))
```

Comment:
generateEvent is also used in Event() with metav1.Now().
if the timestamp is ignored there too, does this mean that there is a bug in generateEvent()?

generateEvent(), calls makeEvent() which always uses the Now() from the recorder clock:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/record/event.go#L328

does this mean that makeEvent() should optionally support a custom timestamp for past events?

Reply:
Great q - yup, currently, the `timestamp` parameter in `generateEvent` is unused: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/record/event.go#L316. If we go the deletion route, we should remove the param from `generateEvent` as well.

I think the central question is whether we ever actually need the ability to generate an event with a custom timestamp? My hypothesis was, given we were seeing so few uses in the code base (and that no one had complained about this bug before), no one actually needs this functionality. But it's very possible my assumption is incorrect :) Do you have thoughts?

Reply:
searching the k/k commit history revealed:
https://github.com/kubernetes/kubernetes/pull/6718

which was added in the v0.17 timeframe. the `timestamp` is not used in the original commit too:
https://github.com/kubernetes/kubernetes/commit/820d0f3e830ed7991f23675369cf3d32ee5c2dc0#diff-1fd310b6d340424c13cbc10ed8945d0aR240-R251

/cc @dchen1107 @vishh 

this code used to exist in `pkg/client/record/event.go`, but is now in client-go, thus is now owned by api-machinery.

as you noted the only consumer of `PastEventf` is the kubelet OOM watcher. it was added in the v1.16 timeframe:
https://github.com/kubernetes/kubernetes/pull/78174

reviewed by you (@mattjmcnaughton :)) and also:
/assign @yujuhong @dashpole 
/cc @gaorong @feiskyer 

> I think the central question is whether we ever actually need the ability to generate an event with a custom timestamp?

indeed.
if i'm not mistaken, past events never worked so this is a valid question.

Reply:
I looked into it a bit, and I vote for removing that option:
- the time difference introduced by using "now" shouldn't be huge
- events are best-effort anyway (and can be silently dropped by the system anyway), so introducing a false impression that we are super accurate with reporting timestamps is misleading in my opinion
- the fact that there was a bug for last 5 years in this code and noone reported it confirms that it's not really needed

So I actually strongly vote for proceeding with this PR and removing this code.

@yastij - FYI

Reply:
I'm in favour of removal for the same reason @wojtek-t is mentioning. Also it's been dropped from the new events implementation too

Reply:
I agree with @wojtek-t's assessment, and am in favor of switching OOM events to use Eventf.

Reply:
Thanks @dashpole 

Given we all seem to be in agreement, I'm approving this PR.